Add support to 32-byte chain work for Bitcoin blocks (#454)
rsksmart/RSKIPs: IPs/RSKIP454.md

---
rskip: 454

> title: Add support to bitcoin blocks with chain work up to 32 unsigned bytes

Author: I made it shorter. Take a look.

created: 25-Oct-24
author: NC
purpose: Sca
layer: Core 
complexity: 1
status: Draft
description: 
---

|RSKIP          | 454                                                                   |
| :------------ |:----------------------------------------------------------------------|
|**Title**      | Add support to bitcoin blocks with chain work up to 32 unsigned bytes |
|**Created**    | 25-Oct-24                                                             |
|**Author**     | NC                                                                    |
|**Purpose**    | Sca                                                                   |
|**Layer**      | Core                                                                  |
|**Complexity** | 1                                                                     |
|**Status**     | Draft                                                                 |

## Abstract

> This RSKIP proposes enabling RSK to accept Bitcoin blocks with chain work up to 32 unsigned bytes(256 bits). This change is necessary to ensure the bridge can process Bitcoin blocks with chain work values that exceed the current 12 unsigned bytes limit(96 bits).

Reviewer: ```suggestion
This RSKIP proposes enabling RSK to accept Bitcoin blocks with chain work up to 32 unsigned bytes (256 bits). This change is necessary to ensure the bridge can process Bitcoin blocks with chain work values that exceed the current 12 unsigned bytes limit (96 bits).
```

> This RSKIP proposes enabling RSK to accept Bitcoin blocks with chain work up to 32 unsigned bytes(256 bits). This change is necessary to ensure the bridge can process Bitcoin blocks with chain work values that exceed the current 12 unsigned bytes limit(96 bits).

Reviewer: ```suggestion
Currently, the bridge can process Bitcoin blocks with chain work values up to 12 unsigned bytes (96 bits).
This RSKIP proposes enabling RSK to increase this limit up to 32 unsigned bytes (256 bits).
```
something like that? wdyt?

Author: I prefer how it is, as my version emphasizes the inevitable necessity of this feature for the future by:
> ... This change is necessary to ensure the bridge can process Bitcoin blocks with chain work values that exceed the current 12 unsigned bytes limit(96 bits).

> Before [RSKIP434](RSKIP434.md) the chain work was even lower (12 signed bytes, 1 bit reserved for the sign and 95 bits used to store the chain work value). This 95 bits limit chain work got surpassed by the block #849,138 causing rsk nodes to be unable to process Bitcoin blocks, resulting in a [Rootstock peg-in / peg-out service outage](https://blog.rootstock.io/noticia/incident-report-rootstock-peg-in-peg-out-service-outage-on-june-24th/).[1]

Reviewer 2: Perhaps RSKIP434 should also be put as a reference?

Author: Done

> Before [RSKIP434](RSKIP435.md) this limitation was even lower, 12 signed bytes, 1 bit reserved for the sign and 95 bits used to store the chain work value. This 95 bits limit chain work got surpassed by the block #849,138 causing rsk nodes to be unable to process Bitcoin blocks, resulting in a [Rootstock peg-in / peg-out service outage](https://blog.rootstock.io/noticia/incident-report-rootstock-peg-in-peg-out-service-outage-on-june-24th/).[1]

Author: Updated

> Before [RSKIP434](RSKIP435.md) this limitation was even lower, 12 signed bytes, 1 bit reserved for the sign and 95 bits used to store the chain work value. This 95 bits limit chain work got surpassed by the block #849,138 causing rsk nodes to be unable to process Bitcoin blocks, resulting in a [Rootstock peg-in / peg-out service outage](https://blog.rootstock.io/noticia/incident-report-rootstock-peg-in-peg-out-service-outage-on-june-24th/).[1]

Reviewer: :)

> Therefore, in order to guarantee the future stability of the Rootstock network, it is necessary to increase the chain work limit before the chain work value exceed the current 12 unsigned byte limit.

Reviewer: ```suggestion
Therefore, in order to guarantee the future stability of the Rootstock network, it is necessary to increase the chain work limit before the chain work value exceed the current 12 unsigned bytes limit.
```

Author: Updated

## Motivation

> Ensure rsk nodes stability and prevent future disruptions in the Rootstock network caused by the inability to process Bitcoin blocks with chain work values that exceed the current 12-byte limit. 

Reviewer: ```suggestion
Ensure rsk nodes stability and prevent future disruptions in the Rootstock network caused by the inability to process Bitcoin blocks with chain work values that exceed the current 12 unsigned bytes limit. 
```

Author: Updated

## Specification

> - Store chain work values of Bitcoin block headers using 32 unsigned bytes. This implies that after this RSKIP gets activated, stored blocks in rsk nodes' storage will be 20 bytes larger in size than before the RSKIP activation.

Reviewer 2: Instead of `20 bytes larger in size than before the RSKIP activation` I'd mention the new size.

Something like "The new block size will go from X to Y bytes in size. N bytes corresponding to chain work value"

Author: Updatede

- Add support to serialize and deserialize chain work values up to 32 unsigned bytes of Bitcoin block headers stored in the Rootstock Bridge.

> - Deprecate binary checkpoint support in benefit of adding support to textual checkpoint format. Checkpoints with 32 bytes chainwork must be in textual format in order to be processed since binary checkpoint format does not support mix bitcoin headers sizes, 12 bytes and 32 bytes chainwork. 

Reviewer 3: ```suggestion
- Deprecate binary checkpoint support in favor of adding support for the textual checkpoint format. Checkpoints with 32-byte chain work must be in textual format to be processed, as the binary checkpoint format does not support mixed Bitcoin header sizes of 12 bytes and 32 bytes for chain work.
```

> - Deprecate binary checkpoint support in benefit of adding support to textual checkpoint format. Checkpoints with 32 bytes chainwork must be in textual format in order to be processed since binary checkpoint format does not support mix bitcoin headers sizes, 12 bytes and 32 bytes chainwork. 

Reviewer: ```suggestion
- Deprecate binary checkpoint support in benefit of adding support to textual checkpoint format. Checkpoints with 32 bytes chainwork must be in textual format in order to be processed since binary checkpoint format does not support mixing bitcoin headers sizes. 
```

Author: Updated

### Backward Compatibility

> This change is a hard fork since we are going to increase by 20 bytes the size for storing block headers in order to support larger chain works(up to 32 unsigned byte). Therefore, all nodes must be updated.

Reviewer 3: ```suggestion
This change is a hard fork, as it increases the size for storing block headers by 20 bytes to support larger chain work values (up to 32 unsigned bytes). Therefore, all nodes must be updated.
```

Author: Updated

> This change is a hard fork since we are going to increase by 20 bytes the size for storing block headers in order to support larger chain works(up to 32 unsigned byte). Therefore, all nodes must be updated.

Reviewer: ```suggestion
This change is a hard fork and therefore all full nodes must be updated.
```

i think this is the general way here, not specifying much

Author: New version is better than first one. Check it out please.

Reviewer: Still, I think we usually just say "this a hard fork" without much explanation. But no big deal

## Rationale

> Given the continued increased in Bitcoin's network difficulty, the chain work value of Bitcoin blocks will continue to grow and will inevitably exceed 96 bits, which would again cause an overflow issue. Therefore, this long-term solution is needed to prevent future disruption in the Rootstock network.

Reviewer 3: Given the continued increase in Bitcoin's network difficulty, the chain work value of Bitcoin blocks will continue to grow and will inevitably exceed 96 bits, which would again cause an overflow issue. Therefore, this long-term solution is needed to prevent future disruption in the Rootstock network.

> Given the continued increased in Bitcoin's network difficulty, the chain work value of Bitcoin blocks will continue to grow and will inevitably exceed 96 bits, which would again cause an overflow issue. Therefore, this long-term solution is needed to prevent future disruption in the Rootstock network.

Reviewer 3: ```suggestion
Given the continued increase in Bitcoin's network difficulty, the chain work value of Bitcoin blocks will continue to grow and will inevitably exceed 96 bits, which would again cause an overflow issue. Therefore, this long-term solution is needed to prevent future disruption in the Rootstock network.
```

Author: Updated

---
rskip: 454
title: Support bitcoin blocks with chain work up to 32 unsigned bytes
created: 25-Oct-24
author: NC
purpose: Sca
layer: Core 
complexity: 1
status: Draft
description: 
---

|RSKIP          | 454                                                                   |
| :------------ |:----------------------------------------------------------------------|
|**Title**      | Support bitcoin blocks with chain work up to 32 unsigned bytes |
|**Created**    | 25-Oct-24                                                             |
|**Author**     | NC                                                                    |
|**Purpose**    | Sca                                                                   |
|**Layer**      | Core                                                                  |
|**Complexity** | 1                                                                     |
|**Status**     | Draft                                                                 |

## Abstract

> This RSKIP proposes enabling RSK to accept Bitcoin blocks with chain work up to 32 unsigned bytes (256 bits). This change is necessary to ensure the bridge can process Bitcoin blocks with chain work values that exceed the current 12 unsigned bytes limit (96 bits).

Reviewer 4: Can you change all RSK references to Rootstock?

### Backward Compatibility

This change is a hard fork, as it increases the size for storing block headers by 20 bytes to support larger chain work values (up to 32 unsigned bytes). Therefore, all nodes must be updated.

## Rationale

Given the continued increase in Bitcoin's network difficulty, the chain work value of Bitcoin blocks will continue to grow and will inevitably exceed 96 bits, which would again cause an overflow issue. Therefore, this long-term solution is needed to prevent future disruption in the Rootstock network. The max chain work value supported now will be 2^256 - 1 or `ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`.

## References

[1] Support bitcoin blocks with chain work up to 32 unsigned bytes [RSKIP434](RSKIP434.md)

[2] Rootstock incident report https://blog.rootstock.io/noticia/incident-report-rootstock-peg-in-peg-out-service-outage-on-june-24th/

[3] Bitcoin block #849,138 https://mempool.space/block/00000000000000000002717c41c14b7865e24af838dbc374fc1952f9cbf7a2ff

> [4] BitcoinJ long-term solution https://github.com/bitcoinj/bitcoinj/pull/3425/, https://github.com/bitcoinj/bitcoinj/pull/3426, https://github.com/bitcoinj/bitcoinj/pull/3427, https://github.com/bitcoinj/bitcoinj/pull/3428/

Reviewer 5: Is our solution the same used as the long-term solution in BitcoinJ ?

Author: Yes, almost. The implementation is very similar, but as Marcos described, we considered that applying the new serialization format only when chain work surpasses 12 bytes will be hard to test. Therefore, we propose using the new serialization format once the RSKIP is activated. This means we will know if this new serialization works as the network upgrade activates.
